File: routes/campaign.js
const express = require("express");
const Campaign = require("../models/Campaign");
const Product = require("../models/Product");
const Promotion = require("../models/Promotion");
const auth = require("../middleware/auth");
const mongoose = require("mongoose");
const router = express.Router();

// Helper function to generate QR code URL (placeholder implementation)
function generateQRCode(campaignId) {
  return `http://16.171.10.235:5000/campaign/${campaignId}`;
}

// POST /api/campaigns - Create a new campaign
router.post("/", auth, async (req, res) => {
  try {
    const {
      name,
      category,
      promotion,
      review,
      products,
      funnelSteps,
      reviewMinimumLength,
      enableSmartFunnel,
      promotionSettings,
      customization,
    } = req.body;

    // Basic validation
    if (!name || !category || !products || products.length === 0) {
      return res.status(400).json({
        message: "Name, category, and at least one product are required",
      });
    }

    // Validate category-specific requirements
    if (category === "promotion" && !promotion) {
      return res.status(400).json({
        message: "Promotion ID is required for promotion campaigns",
      });
    }

    // if (category === "review" && !review) {
    //   return res.status(400).json({
    //     message: "Review configuration is required for review campaigns",
    //   });
    // }

    // Verify products belong to the user
    const userProducts = await Product.find({
      _id: { $in: products },
      seller: req.user.id,
    });

    if (userProducts.length !== products.length) {
      return res.status(400).json({
        message: "One or more products do not belong to you",
      });
    }

    // Create campaign
    const campaign = new Campaign({
      name,
      category,
      seller: req.user.id,
      promotion: category === "promotion" ? promotion : undefined,
      review: category === "review" ? review : undefined,
      products,
      funnelSteps: funnelSteps || [],
      reviewMinimumLength: reviewMinimumLength || 10,
      enableSmartFunnel: enableSmartFunnel || false,
      promotionSettings:
        category === "promotion" ? promotionSettings : undefined,
      customization: customization || {},
    });

    await campaign.save();

    // Generate QR code URL
    campaign.qrCodeUrl = generateQRCode(campaign._id);
    await campaign.save();

    // Populate references before sending response
    await campaign.populate([
      { path: "seller", select: "name email" },
      { path: "products", select: "name marketplace" },
      { path: "promotion", select: "name type description" },
    ]);

    res.status(201).json({
      message: "Campaign created successfully",
      campaign,
    });
  } catch (error) {
    console.error("Create campaign error:", error);
    res.status(500).json({ message: "Server error" });
  }
});

// GET /api/campaigns - Get all campaigns for the current user
router.get("/", auth, async (req, res) => {
  try {
    const { status, category, page = 1, limit = 10 } = req.query;

    // Build filter object
    const filter = { seller: req.user.id };

    if (status && ["active", "paused", "ended"].includes(status)) {
      filter.status = status;
    }

    if (category && ["review", "promotion"].includes(category)) {
      filter.category = category;
    }

    const skip = (page - 1) * limit;

    const campaigns = await Campaign.find(filter)
      .populate([
        { path: "seller", select: "name email" },
        { path: "products", select: "name marketplace" },
        { path: "promotion", select: "name type description" },
      ])
      .sort({ createdAt: -1 })
      .limit(limit * 1)
      .skip(skip);

    const total = await Campaign.countDocuments(filter);

    res.json({
      campaigns,
      pagination: {
        current: page,
        pages: Math.ceil(total / limit),
        total,
      },
    });
  } catch (error) {
    console.error("Get campaigns error:", error);
    res.status(500).json({ message: "Server error" });
  }
});

// GET /api/campaigns/:id - Get a specific campaign by ID
router.get("/:id", auth, async (req, res) => {
  try {
    const campaign = await Campaign.findOne({
      _id: req.params.id,
      seller: req.user.id,
    }).populate([
      { path: "seller", select: "name email" },
      { path: "products", select: "name marketplace marketplaceProductId" },
      { path: "promotion", select: "name type description provider value" },
    ]);

    if (!campaign) {
      return res.status(404).json({ message: "Campaign not found" });
    }

    res.json(campaign);
  } catch (error) {
    console.error("Get campaign error:", error);
    if (error.name === "CastError") {
      return res.status(400).json({ message: "Invalid campaign ID" });
    }
    res.status(500).json({ message: "Server error" });
  }
});

// PUT /api/campaigns/:id - Update a campaign
router.put("/:id", auth, async (req, res) => {
  try {
    const {
      name,
      status,
      products,
      funnelSteps,
      reviewMinimumLength,
      enableSmartFunnel,
      promotionSettings,
      customization,
    } = req.body;

    const campaign = await Campaign.findOne({
      _id: req.params.id,
      seller: req.user.id,
    });

    if (!campaign) {
      return res.status(404).json({ message: "Campaign not found" });
    }

    // Validate status if provided
    if (status && !["active", "paused", "ended"].includes(status)) {
      return res.status(400).json({ message: "Invalid status" });
    }

    // Verify products belong to the user if products are being updated
    if (products && products.length > 0) {
      const userProducts = await Product.find({
        _id: { $in: products },
        seller: req.user.id,
      });

      if (userProducts.length !== products.length) {
        return res.status(400).json({
          message: "One or more products do not belong to you",
        });
      }
    }

    // Update fields
    if (name) campaign.name = name;
    if (status) campaign.status = status;
    if (products) campaign.products = products;
    if (funnelSteps) campaign.funnelSteps = funnelSteps;
    if (reviewMinimumLength !== undefined)
      campaign.reviewMinimumLength = reviewMinimumLength;
    if (enableSmartFunnel !== undefined)
      campaign.enableSmartFunnel = enableSmartFunnel;
    if (promotionSettings)
      campaign.promotionSettings = {
        ...campaign.promotionSettings,
        ...promotionSettings,
      };
    if (customization)
      campaign.customization = { ...campaign.customization, ...customization };

    await campaign.save();

    // Populate before sending response
    await campaign.populate([
      { path: "seller", select: "name email" },
      { path: "products", select: "name marketplace" },
      { path: "promotion", select: "name type description" },
    ]);

    res.json({
      message: "Campaign updated successfully",
      campaign,
    });
  } catch (error) {
    console.error("Update campaign error:", error);
    if (error.name === "CastError") {
      return res.status(400).json({ message: "Invalid campaign ID" });
    }
    res.status(500).json({ message: "Server error" });
  }
});

// DELETE /api/campaigns/:id - change status

//update status if active make it ended  and if ended than active

router.delete("/:id", auth, async (req, res) => {
  try {
    const campaign = await Campaign.findOne({
      _id: req.params.id,
      seller: req.user.id,
    });

    if (!campaign) {
      return res.status(404).json({ message: "Campaign not found" });
    }

    //update status if active make it ended  and if ended than active
    campaign.status = campaign.status === "active" ? "ended" : "active";
    await campaign.save();

    // await Campaign.findByIdAndDelete(req.params.id);

    res.json({ message: "Campaign deleted successfully" });
  } catch (error) {
    console.error("Delete campaign error:", error);
    if (error.name === "CastError") {
      return res.status(400).json({ message: "Invalid campaign ID" });
    }
    res.status(500).json({ message: "Server error" });
  }
});

// GET /api/campaigns/stats/dashboard - Get campaign statistics
router.get("/stats/dashboard", auth, async (req, res) => {
  try {
    const totalCampaigns = await Campaign.countDocuments({
      seller: req.user.id,
    });
    const activeCampaigns = await Campaign.countDocuments({
      seller: req.user.id,
      status: "active",
    });
    const pausedCampaigns = await Campaign.countDocuments({
      seller: req.user.id,
      status: "paused",
    });

    // Count by category
    const campaignsByCategory = await Campaign.aggregate([
      { $match: { seller: new mongoose.Types.ObjectId(req.user.id) } },
      { $group: { _id: "$category", count: { $sum: 1 } } },
    ]);

    // Analytics summary
    const analyticsData = await Campaign.aggregate([
      { $match: { seller: new mongoose.Types.ObjectId(req.user.id) } },
      {
        $group: {
          _id: null,
          totalScans: { $sum: "$analytics.totalScans" },
          totalCompletions: { $sum: "$analytics.totalCompletions" },
          totalRedemptions: { $sum: "$analytics.totalRedemptions" },
        },
      },
    ]);

    const analytics = analyticsData[0] || {
      totalScans: 0,
      totalCompletions: 0,
      totalRedemptions: 0,
    };

    res.json({
      summary: {
        total: totalCampaigns,
        active: activeCampaigns,
        paused: pausedCampaigns,
        ended: totalCampaigns - activeCampaigns - pausedCampaigns,
      },
      byCategory: campaignsByCategory,
      analytics,
    });
  } catch (error) {
    console.error("Get campaign stats error:", error);
    res.status(500).json({ message: "Server error" });
  }
});

// PUT /api/campaigns/:id/analytics - Update campaign analytics (for tracking scans/completions)
router.put("/:id/analytics", auth, async (req, res) => {
  try {
    const { scans, completions, redemptions } = req.body;

    const campaign = await Campaign.findOne({
      _id: req.params.id,
      seller: req.user.id,
    });

    if (!campaign) {
      return res.status(404).json({ message: "Campaign not found" });
    }

    // Update analytics
    if (scans !== undefined) campaign.analytics.totalScans += scans;
    if (completions !== undefined)
      campaign.analytics.totalCompletions += completions;
    if (redemptions !== undefined)
      campaign.analytics.totalRedemptions += redemptions;

    // Recalculate conversion rate
    if (campaign.analytics.totalScans > 0) {
      campaign.analytics.conversionRate =
        (campaign.analytics.totalCompletions / campaign.analytics.totalScans) *
        100;
    }

    await campaign.save();

    res.json({
      message: "Analytics updated successfully",
      analytics: campaign.analytics,
    });
  } catch (error) {
    console.error("Update analytics error:", error);
    res.status(500).json({ message: "Server error" });
  }
});

module.exports = router;

File: yourapp-go/controllers/campaign.go
package controllers

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"yourapp/config"
	"yourapp/middleware"
	"yourapp/models"
)

// CreateCampaignRequest represents the request body for creating a campaign
type CreateCampaignRequest struct {
	Name                string                   `json:"name" binding:"required"`
	Category            string                   `json:"category" binding:"required"`
	Promotion           string                   `json:"promotion,omitempty"`
	Review              string                   `json:"review,omitempty"`
	Products            []string                 `json:"products" binding:"required,min=1"`
	FunnelSteps         []models.FunnelStep      `json:"funnelSteps"`
	ReviewMinimumLength int                      `json:"reviewMinimumLength"`
	EnableSmartFunnel   bool                     `json:"enableSmartFunnel"`
	PromotionSettings   models.PromotionSettings `json:"promotionSettings"`
	Customization       models.Customization     `json:"customization"`
}

// UpdateCampaignRequest represents the request body for updating a campaign
type UpdateCampaignRequest struct {
	Name                *string                   `json:"name,omitempty"`
	Status              *string                   `json:"status,omitempty"`
	Products            []string                  `json:"products,omitempty"`
	FunnelSteps         []models.FunnelStep       `json:"funnelSteps,omitempty"`
	ReviewMinimumLength *int                      `json:"reviewMinimumLength,omitempty"`
	EnableSmartFunnel   *bool                     `json:"enableSmartFunnel,omitempty"`
	PromotionSettings   *models.PromotionSettings `json:"promotionSettings,omitempty"`
	Customization       *models.Customization     `json:"customization,omitempty"`
}

// UpdateAnalyticsRequest represents the request body for updating analytics
type UpdateAnalyticsRequest struct {
	Scans       *int `json:"scans,omitempty"`
	Completions *int `json:"completions,omitempty"`
	Redemptions *int `json:"redemptions,omitempty"`
}

// GenerateQRCode generates a QR code URL for a campaign
func GenerateQRCode(campaignID primitive.ObjectID) string {
	return fmt.Sprintf("http://16.171.10.235:5000/campaign/%s", campaignID.Hex())
}

// CreateCampaign creates a new campaign
func CreateCampaign(c *gin.Context) {
	var req CreateCampaignRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		log.Printf("Validation error: %v", err)
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Invalid request data",
		})
		return
	}

	// Validate category
	if req.Category != "promotion" && req.Category != "review" {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Category must be either 'promotion' or 'review'",
		})
		return
	}

	// Validate category-specific requirements
	if req.Category == "promotion" && req.Promotion == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Promotion ID is required for promotion campaigns",
		})
		return
	}

	// Get user ID from context
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Convert product IDs to ObjectIDs
	productIDs := make([]primitive.ObjectID, len(req.Products))
	for i, pid := range req.Products {
		objID, err := primitive.ObjectIDFromHex(pid)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid product ID format"})
			return
		}
		productIDs[i] = objID
	}

	// Verify products belong to the user
	productCount, err := db.Collection("products").CountDocuments(ctx, bson.M{
		"_id":    bson.M{"$in": productIDs},
		"seller": userID,
	})
	if err != nil {
		log.Printf("Database error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		return
	}

	if int(productCount) != len(productIDs) {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "One or more products do not belong to you",
		})
		return
	}

	// Create new campaign with defaults
	campaign := models.NewCampaign()
	campaign.Name = req.Name
	campaign.Category = req.Category
	campaign.Seller = userID
	campaign.Products = productIDs

	// Set optional fields
	if req.ReviewMinimumLength > 0 {
		campaign.ReviewMinimumLength = req.ReviewMinimumLength
	}
	campaign.EnableSmartFunnel = req.EnableSmartFunnel

	// Merge customization settings
	if req.Customization.PrimaryColor != "" {
		campaign.Customization.PrimaryColor = req.Customization.PrimaryColor
	}
	if req.Customization.LogoURL != "" {
		campaign.Customization.LogoURL = req.Customization.LogoURL
	}
	if req.Customization.BackgroundStyle != "" {
		campaign.Customization.BackgroundStyle = req.Customization.BackgroundStyle
	}
	if req.Customization.CustomMessage != "" {
		campaign.Customization.CustomMessage = req.Customization.CustomMessage
	}

	// Merge promotion settings
	if req.PromotionSettings.CodeType != "" {
		campaign.PromotionSettings.CodeType = req.PromotionSettings.CodeType
	}
	if req.PromotionSettings.CodeValue != "" {
		campaign.PromotionSettings.CodeValue = req.PromotionSettings.CodeValue
	}
	if req.PromotionSettings.DeliveryType != "" {
		campaign.PromotionSettings.DeliveryType = req.PromotionSettings.DeliveryType
	}
	if req.PromotionSettings.MaxRedemptions > 0 {
		campaign.PromotionSettings.MaxRedemptions = req.PromotionSettings.MaxRedemptions
	}
	if req.PromotionSettings.ExpiryDate != nil {
		campaign.PromotionSettings.ExpiryDate = req.PromotionSettings.ExpiryDate
	}

	// Set funnel steps
	if req.FunnelSteps != nil {
		campaign.FunnelSteps = req.FunnelSteps
	}

	// Handle optional promotion/review IDs
	if req.Category == "promotion" && req.Promotion != "" {
		promotionID, err := primitive.ObjectIDFromHex(req.Promotion)
		if err == nil {
			campaign.Promotion = &promotionID
		}
	}

	if req.Category == "review" && req.Review != "" {
		reviewID, err := primitive.ObjectIDFromHex(req.Review)
		if err == nil {
			campaign.Review = &reviewID
		}
	}

	// Insert campaign
	result, err := db.Collection(campaign.CollectionName()).InsertOne(ctx, campaign)
	if err != nil {
		log.Printf("Database insert error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to create campaign"})
		return
	}

	campaignID := result.InsertedID.(primitive.ObjectID)

	// Generate QR code URL and update
	qrCodeURL := GenerateQRCode(campaignID)
	_, err = db.Collection(campaign.CollectionName()).UpdateOne(
		ctx,
		bson.M{"_id": campaignID},
		bson.M{"$set": bson.M{"qrCodeUrl": qrCodeURL}},
	)
	if err != nil {
		log.Printf("Failed to update QR code: %v", err)
	}

	campaign.ID = campaignID
	campaign.QRCodeURL = qrCodeURL

	c.JSON(http.StatusCreated, gin.H{
		"message":  "Campaign created successfully",
		"campaign": campaign,
	})
}

// GetCampaigns retrieves all campaigns for the current user
func GetCampaigns(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Query parameters
	status := c.Query("status")
	category := c.Query("category")
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "10"))

	// Build filter
	filter := bson.M{"seller": userID}

	if status != "" && (status == "active" || status == "paused" || status == "ended") {
		filter["status"] = status
	}

	if category != "" && (category == "review" || category == "promotion") {
		filter["category"] = category
	}

	skip := int64((page - 1) * limit)

	// Find campaigns
	findOptions := options.Find().
		SetSort(bson.D{{Key: "createdAt", Value: -1}}).
		SetLimit(int64(limit)).
		SetSkip(skip)

	cursor, err := db.Collection("campaigns").Find(ctx, filter, findOptions)
	if err != nil {
		log.Printf("Database error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		return
	}
	defer cursor.Close(ctx)

	var campaigns []models.Campaign
	if err := cursor.All(ctx, &campaigns); err != nil {
		log.Printf("Error decoding campaigns: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Error decoding campaigns"})
		return
	}

	// Initialize empty array if nil
	if campaigns == nil {
		campaigns = []models.Campaign{}
	}

	// Count total
	total, err := db.Collection("campaigns").CountDocuments(ctx, filter)
	if err != nil {
		log.Printf("Error counting documents: %v", err)
		total = 0
	}

	c.JSON(http.StatusOK, gin.H{
		"campaigns": campaigns,
		"pagination": gin.H{
			"current": page,
			"pages":   (total + int64(limit) - 1) / int64(limit),
			"total":   total,
		},
	})
}

// GetCampaignByID retrieves a specific campaign by ID
func GetCampaignByID(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	campaignID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid campaign ID"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	var campaign models.Campaign
	err = db.Collection("campaigns").FindOne(ctx, bson.M{
		"_id":    campaignID,
		"seller": userID,
	}).Decode(&campaign)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"message": "Campaign not found"})
		} else {
			log.Printf("Database error: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		}
		return
	}

	c.JSON(http.StatusOK, campaign)
}

// UpdateCampaign updates a campaign
func UpdateCampaign(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	campaignID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid campaign ID"})
		return
	}

	var req UpdateCampaignRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid request data"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Validate status
	if req.Status != nil && *req.Status != "active" && *req.Status != "paused" && *req.Status != "ended" {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid status"})
		return
	}

	// Build update document
	update := bson.M{
		"$set": bson.M{
			"updatedAt": time.Now(),
		},
	}

	if req.Name != nil {
		update["$set"].(bson.M)["name"] = *req.Name
	}
	if req.Status != nil {
		update["$set"].(bson.M)["status"] = *req.Status
	}
	if req.FunnelSteps != nil {
		update["$set"].(bson.M)["funnelSteps"] = req.FunnelSteps
	}
	if req.ReviewMinimumLength != nil {
		update["$set"].(bson.M)["reviewMinimumLength"] = *req.ReviewMinimumLength
	}
	if req.EnableSmartFunnel != nil {
		update["$set"].(bson.M)["enableSmartFunnel"] = *req.EnableSmartFunnel
	}
	if req.PromotionSettings != nil {
		update["$set"].(bson.M)["promotionSettings"] = req.PromotionSettings
	}
	if req.Customization != nil {
		update["$set"].(bson.M)["customization"] = req.Customization
	}

	// Handle products update
	if req.Products != nil && len(req.Products) > 0 {
		productIDs := make([]primitive.ObjectID, len(req.Products))
		for i, pid := range req.Products {
			objID, err := primitive.ObjectIDFromHex(pid)
			if err != nil {
				c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid product ID format"})
				return
			}
			productIDs[i] = objID
		}

		// Verify products
		productCount, err := db.Collection("products").CountDocuments(ctx, bson.M{
			"_id":    bson.M{"$in": productIDs},
			"seller": userID,
		})
		if err != nil || int(productCount) != len(productIDs) {
			c.JSON(http.StatusBadRequest, gin.H{"message": "One or more products do not belong to you"})
			return
		}

		update["$set"].(bson.M)["products"] = productIDs
	}

	// Update campaign
	result, err := db.Collection("campaigns").UpdateOne(
		ctx,
		bson.M{"_id": campaignID, "seller": userID},
		update,
	)

	if err != nil {
		log.Printf("Database update error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to update campaign"})
		return
	}

	if result.MatchedCount == 0 {
		c.JSON(http.StatusNotFound, gin.H{"message": "Campaign not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Campaign updated successfully"})
}

// DeleteCampaign toggles campaign status between active and ended
func DeleteCampaign(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	campaignID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid campaign ID"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Get current campaign
	var campaign models.Campaign
	err = db.Collection("campaigns").FindOne(ctx, bson.M{
		"_id":    campaignID,
		"seller": userID,
	}).Decode(&campaign)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"message": "Campaign not found"})
		} else {
			log.Printf("Database error: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		}
		return
	}

	// Use the model method to toggle status
	campaign.ToggleStatus()

	_, err = db.Collection("campaigns").UpdateOne(
		ctx,
		bson.M{"_id": campaignID},
		bson.M{"$set": bson.M{"status": campaign.Status, "updatedAt": campaign.UpdatedAt}},
	)

	if err != nil {
		log.Printf("Database update error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to update campaign"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Campaign status updated successfully"})
}

// GetCampaignStats retrieves campaign statistics for dashboard
func GetCampaignStats(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Count campaigns by status
	total, _ := db.Collection("campaigns").CountDocuments(ctx, bson.M{"seller": userID})
	active, _ := db.Collection("campaigns").CountDocuments(ctx, bson.M{"seller": userID, "status": "active"})
	paused, _ := db.Collection("campaigns").CountDocuments(ctx, bson.M{"seller": userID, "status": "paused"})
	ended := total - active - paused

	// Campaigns by category aggregation
	categoryPipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.D{{Key: "seller", Value: userID}}}},
		{{Key: "$group", Value: bson.D{
			{Key: "_id", Value: "$category"},
			{Key: "count", Value: bson.D{{Key: "$sum", Value: 1}}},
		}}},
	}

	categoryCursor, err := db.Collection("campaigns").Aggregate(ctx, categoryPipeline)
	if err != nil {
		log.Printf("Aggregation error: %v", err)
	}

	var categoryResults []bson.M
	if err == nil {
		categoryCursor.All(ctx, &categoryResults)
	}
	if categoryResults == nil {
		categoryResults = []bson.M{}
	}

	// Analytics summary aggregation
	analyticsPipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.D{{Key: "seller", Value: userID}}}},
		{{Key: "$group", Value: bson.D{
			{Key: "_id", Value: nil},
			{Key: "totalScans", Value: bson.D{{Key: "$sum", Value: "$analytics.totalScans"}}},
			{Key: "totalCompletions", Value: bson.D{{Key: "$sum", Value: "$analytics.totalCompletions"}}},
			{Key: "totalRedemptions", Value: bson.D{{Key: "$sum", Value: "$analytics.totalRedemptions"}}},
		}}},
	}

	analyticsCursor, err := db.Collection("campaigns").Aggregate(ctx, analyticsPipeline)
	analytics := gin.H{
		"totalScans":       0,
		"totalCompletions": 0,
		"totalRedemptions": 0,
	}

	if err == nil && analyticsCursor.Next(ctx) {
		var result bson.M
		analyticsCursor.Decode(&result)
		if result != nil {
			analytics = gin.H(result)
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"summary": gin.H{
			"total":  total,
			"active": active,
			"paused": paused,
			"ended":  ended,
		},
		"byCategory": categoryResults,
		"analytics":  analytics,
	})
}

// UpdateCampaignAnalytics updates campaign analytics
func UpdateCampaignAnalytics(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	campaignID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid campaign ID"})
		return
	}

	var req UpdateAnalyticsRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid request data"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Get current campaign
	var campaign models.Campaign
	err = db.Collection("campaigns").FindOne(ctx, bson.M{
		"_id":    campaignID,
		"seller": userID,
	}).Decode(&campaign)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"message": "Campaign not found"})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		}
		return
	}

	// Update analytics
	update := bson.M{}
	if req.Scans != nil {
		campaign.Analytics.TotalScans += *req.Scans
		update["analytics.totalScans"] = campaign.Analytics.TotalScans
	}
	if req.Completions != nil {
		campaign.Analytics.TotalCompletions += *req.Completions
		update["analytics.totalCompletions"] = campaign.Analytics.TotalCompletions
	}
	if req.Redemptions != nil {
		campaign.Analytics.TotalRedemptions += *req.Redemptions
		update["analytics.totalRedemptions"] = campaign.Analytics.TotalRedemptions
	}

	// Use model method to calculate conversion rate
	campaign.CalculateConversionRate()
	update["analytics.conversionRate"] = campaign.Analytics.ConversionRate
	update["updatedAt"] = time.Now()

	_, err = db.Collection("campaigns").UpdateOne(
		ctx,
		bson.M{"_id": campaignID},
		bson.M{"$set": update},
	)

	if err != nil {
		log.Printf("Database update error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to update analytics"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":   "Analytics updated successfully",
		"analytics": campaign.Analytics,
	})
}
